src/ephy-sync-crypto.c

```diff
@@ -0,0 +1,952 @@
+                                     const char *dlg,
+                                     const char *ext,
+                                     const char *hash,
+                                     const char *host,
+                                     const char *method,
+                                     const char *nonce,
+                                     guint       port,
+                                     const char *resource,
+                                     gint64      ts)
+{
+  EphySyncCryptoHawkArtifacts *artifacts;
+
+  artifacts = g_slice_new (EphySyncCryptoHawkArtifacts);
+  artifacts->app = g_strdup (app);
+  artifacts->dlg = g_strdup (dlg);
+  artifacts->ext = g_strdup (ext);
+  artifacts->hash = g_strdup (hash);
+  artifacts->host = g_strdup (host);
+  artifacts->method = g_strdup (method);
+  artifacts->nonce = g_strdup (nonce);
+  artifacts->port = g_strdup_printf ("%u", port);
+  artifacts->resource = g_strdup (resource);
+  artifacts->ts = g_strdup_printf ("%ld", ts);
+
+  return artifacts;
+}
+
+static EphySyncCryptoHawkHeader *
+ephy_sync_crypto_hawk_header_new (char                        *header,
+                                  EphySyncCryptoHawkArtifacts *artifacts)
+{
+  EphySyncCryptoHawkHeader *hheader;
+
+  hheader = g_slice_new (EphySyncCryptoHawkHeader);
+  hheader->header = header;
+  hheader->artifacts = artifacts;
+
+  return hheader;
+}
+
+static EphySyncCryptoRSAKeyPair *
+ephy_sync_crypto_rsa_key_pair_new (struct rsa_public_key  public,
+                                   struct rsa_private_key private)
+{
+  EphySyncCryptoRSAKeyPair *keypair;
+
+  keypair = g_slice_new (EphySyncCryptoRSAKeyPair);
+  keypair->public = public;
+  keypair->private = private;
+
+  return keypair;
+}
+
+void
+ephy_sync_crypto_hawk_options_free (EphySyncCryptoHawkOptions *options)
+{
+  g_return_if_fail (options != NULL);
+
+  g_free (options->app);
+  g_free (options->dlg);
+  g_free (options->ext);
+  g_free (options->content_type);
+  g_free (options->hash);
+  g_free (options->local_time_offset);
+  g_free (options->nonce);
+  g_free (options->payload);
+  g_free (options->timestamp);
+
+  g_slice_free (EphySyncCryptoHawkOptions, options);
+}
+
+static void
+ephy_sync_crypto_hawk_artifacts_free (EphySyncCryptoHawkArtifacts *artifacts)
+{
+  g_return_if_fail (artifacts != NULL);
+
+  g_free (artifacts->app);
+  g_free (artifacts->dlg);
+  g_free (artifacts->ext);
+  g_free (artifacts->hash);
+  g_free (artifacts->host);
+  g_free (artifacts->method);
+  g_free (artifacts->nonce);
+  g_free (artifacts->port);
+  g_free (artifacts->resource);
+  g_free (artifacts->ts);
+
+  g_slice_free (EphySyncCryptoHawkArtifacts, artifacts);
+}
+
+void
+ephy_sync_crypto_hawk_header_free (EphySyncCryptoHawkHeader *hheader)
+{
+  g_return_if_fail (hheader != NULL);
+
+  g_free (hheader->header);
+  ephy_sync_crypto_hawk_artifacts_free (hheader->artifacts);
+
+  g_slice_free (EphySyncCryptoHawkHeader, hheader);
+}
+
+void
```

Comment:
@mcatanzaro The reason why `ephy_sync_crypto_hawk_header_free` and `ephy_sync_crypto_rsa_key_pair_free` are public but their corresponding new functions are not, is that both the `HawkHeader` and the `RSAKeypair` are generated in `ephy-sync-crypto` but they are used in `ephy-sync-service`. So I had to free them in `ephy-sync-service` after they were no longer used, that's why I made the free functions non-static.